packages/indicators/src/AnalyticsRepository.ts

```diff
@@ -0,0 +1,128 @@
+  expandedFetchOptions: FetchOptions;
+}
+
+/**
+ * This class encapsulates analytics fetching and retrieval of fetched data. The motivation
+ * behind it is to optimise data fetches by minimising their count.
+ *
+ * An example:
+ * * Indicator A wants to fetch FINAL_EACH_YEAR data for element BCD
+ * * Indicator B wants to fetch FINAL_EACH_YEAR for element BCD and a year offset of -1
+ * ("get me the results for the previous year")
+ * * The original date range is 2018 - 2020
+ *
+ * We can combine both fetches in one, and fetch BCD for years 2017 - 2020, so that we can include
+ * -1 offset data for the 2018 boundary.
+ *
+ * This means that the "scope" of the fetched data will be wider than the the original.
+ * This class hides the scope transformation and allows each client (indicator) to retrieve
+ * correct data based on the original scope (`fetchOptions`).
+ */
+export class AnalyticsRepository {
+  private aggregator: Aggregator;
+
+  private populatedData: Data | null = null;
+
+  constructor(aggregator: Aggregator) {
+    this.aggregator = aggregator;
+  }
+
+  private get data(): Data {
+    if (!this.populatedData) {
+      throw new Error('Please run fetchAnalytics first!');
+    }
+    return this.populatedData;
+  }
+
+  private set data(data: Data) {
+    this.populatedData = data;
+  }
+
+  isPopulated() {
+    return !!this.populatedData;
+  }
+
+  /**
+   * @param allAggregations A combined list of aggregations across all data elements
+   */
+  async populate(
+    dataElementCodes: string[],
+    allAggregations: Aggregation[],
+    fetchOptions: FetchOptions,
+  ) {
+    const expandedFetchOptions = expandFetchOptionsToSpanAggregations(
+      fetchOptions,
+      allAggregations,
+    );
+    const { results: analytics } = await this.aggregator.fetchAnalytics(
+      dataElementCodes,
+      expandedFetchOptions,
+    );
+
+    this.data = {
+      analyticsByElement: groupBy(analytics, 'dataElement'),
+      fetchOptions,
+      expandedFetchOptions,
+    };
+  }
+
+  getAnalyticsForDataElement = (dataElement: string) =>
+    this.data.analyticsByElement[dataElement] || [];
+
+  /**
+   * This method should be used for analytics in root indicators. The original fetch options are used,
+   * to restrict the top-level data in the originally requested dimensions
+   */
+  aggregateRootAnalytics = (analytics: Analytic[], aggregations: Aggregation[]) => {
+    const analyticsMatchingFetchOptions = this.keepAnalyticsMatchingFetchOptions(
+      analytics,
+      this.data.fetchOptions,
+    );
+    return this.aggregateAnalytics(
+      analyticsMatchingFetchOptions,
+      aggregations,
+      this.data.fetchOptions,
+    );
+  };
+
+  /**
+   * This method should be used for analytics in nested indicators.  Expanded fetch options are used,
+   * to allow calculations that may require data outside the originally requested dimensions
+   */
+  aggregateNestedAnalytics = (analytics: Analytic[], aggregations: Aggregation[]) =>
```

Comment:
It really took a while for me to get my head around what aggregateRootAnalytics vs. aggregateNestedAnalytics meant. I don't really think there are better names, but if you have any thoughts in improving dev understanding here, let me know

Reply:
Yes I believe that's the weakest point of my implementation... But it's the best version I ended up with.

So the comments above them didn't help? Should we just accept that the shrink/expand of fetchOptions (and resulting need for a different nested/root logic) is just complicated business, or should we brainstorm ways to make this even simpler?

Reply:
The comments helped, but didn't get me all the way there. I think it might be nice to illustrate the flow somehow, because I struggled to understand the full flow. Tbh I still do - here's my current understanding:
- nested analytics will already have been aggregated further down the chain by their own indicator builder
- that indicator builder will have aggregated them using the dates, as when it aggregates its own analytics it will do so using the root aggregation process
- at this point the root indicator builder is applying another layer of aggregation on top of the nested analytics that have been passed up, and shouldn't apply the tight date constraints as the dates will have been sorted out by the nested builders

Even in writing that I'm not sure it's true. Question:
- Is it the case that each indicator builder will be considered the "root" at some point, or is there just one root per fetch?

Reply:
There is only one root per fetch